File: prototype/rails/src/app/domain/shop/shop_name.rb
module Domain
  module Shop
    module ShopName
      module_function

      # A seller signs up with an address and names their shop later, so the
      # storefront falls back to the part of the address in front of the host.
      def of(shop_name:, email:)
        named = shop_name.to_s.strip

        named.empty? ? email.to_s.split("@").first.to_s : named
      end
    end
  end
end

File: prototype/node/src/app/core/shop/shop-name.ts
/**
 * How a seller reads on the storefront. A seller signs up with an address and
 * names their shop later, so an unnamed shop falls back to the part of the
 * address in front of the host.
 */
export function shopName(seller: { shopName: string | null; email: string }): string {
  const named = (seller.shopName ?? '').trim()

  return named.length === 0 ? (seller.email.split('@')[0] ?? '') : named
}
